File: src/main/java/com/gazman/db/DBThread.java
package com.gazman.db;

import android.os.Handler;
import android.os.Looper;

/**
 * Created by Ilya Gazman on 7/9/2016.
 */
public final class DBThread {

    static Handler handler;

    static {
        new Thread(new Runnable() {
            @Override
            public void run() {
                Looper.prepare();
                handler = new Handler(Looper.myLooper());
                Looper.loop();
            }
        }).start();
    }

    public static void execute(Runnable runnable) {
        while (handler == null) {
            Thread.yield();
        }

        if (isOnDbThread()) {
            runnable.run();
        } else {
            handler.post(runnable);
        }
    }

    public static void synchronize() {
        synchronized (DBThread.class) {
            execute(() -> {
                synchronized (DBThread.class) {
                    DBThread.class.notifyAll();
                }
            });
            try {
                DBThread.class.wait();
            } catch (InterruptedException e) {
                ErrorAnalytics.logException(e);
                e.printStackTrace();
            }
        }
    }

    public static boolean isOnDbThread() {
        return handler.getLooper() == Looper.myLooper();
    }
}

File: src/main/java/com/gazman/db/DBThread.kt
package com.gazman.db

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Created by Ilya Gazman on 7/9/2016.
 */
object DBThread {
    var handler = Handler(HandlerThread("db_thread").looper)
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()

    fun execute(runnable: Runnable) {
        if (isOnDbThread) {
            runnable.run()
        } else {
            handler.post(runnable)
        }
    }

    fun synchronize() {
        lock.withLock {
            execute {
                lock.withLock {
                    condition.signalAll()
                }
            }
            condition.await()
        }
    }

    private val isOnDbThread: Boolean
        get() = handler.looper == Looper.myLooper()
}